File: src/api/qoder.ts
import { invokeWithResponseSchema, listenWithResponseSchema } from "@/api/ipc";
import {
	type AgentRuntimeConfig,
	CompiledAgentLoginStatusSchema,
	CompiledAgentRunResultSchema,
	CompiledAgentRuntimeConfigSchema,
	CompiledAgentRuntimeStatusSchema,
} from "@/types/agent";

/** Checks the local Qoder CLI account state through its documented status command. */
const checkQoderLogin = () =>
	invokeWithResponseSchema("check_qoder_login", CompiledAgentLoginStatusSchema);

/** Returns the complete Qoder CLI status needed for the first render. */
const checkQoderInitStatus = () =>
	invokeWithResponseSchema(
		"check_qoder_init_status",
		CompiledAgentRuntimeStatusSchema,
	);

/** Returns Qoder's runtime model override when one is available. */
const getQoderRuntimeConfig = () =>
	invokeWithResponseSchema(
		"get_qoder_runtime_config",
		CompiledAgentRuntimeConfigSchema,
	);

/** Subscribes to future native Qoder runtime configuration updates. */
const onQoderConfigChanged = (listener: (config: AgentRuntimeConfig) => void) =>
	listenWithResponseSchema(
		"qoder-config-changed",
		CompiledAgentRuntimeConfigSchema,
		listener,
	);

/** Sends one task through Qoder CLI's non-interactive stream-json mode. */
const runQoderTask = (query: string) =>
	invokeWithResponseSchema("run_qoder_task", CompiledAgentRunResultSchema, {
		request: { query },
	});

export {
	checkQoderInitStatus,
	checkQoderLogin,
	getQoderRuntimeConfig,
	onQoderConfigChanged,
	runQoderTask,
};

File: src-tauri/src/commands/qoder.rs
use crate::adapters::qoder::SystemQoderAdapter;
use crate::dto::agent::{
    AgentInitStatusResponse, AgentLoginStatusResponse, AgentRunResponse, AgentRuntimeConfigResponse,
};
use crate::error::{AppError, IpcError};
use crate::services::agent::{
    check_agent_init_status, check_agent_login, load_agent_runtime_config, run_agent_task,
};
use serde::Deserialize;

/// User input accepted by the Qoder task command.
#[derive(Debug, Clone, PartialEq, Eq, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RunQoderTaskRequest {
    /// Natural-language task sent to Qoder's non-interactive runner.
    query: String,
}

/// Checks whether a locally installed Qoder CLI has an active account.
#[tauri::command]
pub async fn check_qoder_login() -> Result<AgentLoginStatusResponse, IpcError> {
    tauri::async_runtime::spawn_blocking(|| check_agent_login(&SystemQoderAdapter))
        .await
        .map_err(|_| IpcError::from(AppError::WorkerFailed))?
        .map(Into::into)
        .map_err(Into::into)
}

/// Returns the complete first-load Qoder status from independent probes.
#[tauri::command]
pub async fn check_qoder_init_status() -> Result<AgentInitStatusResponse, IpcError> {
    tauri::async_runtime::spawn_blocking(|| check_agent_init_status(&SystemQoderAdapter))
        .await
        .map_err(|_| IpcError::from(AppError::WorkerFailed))?
        .map(Into::into)
        .map_err(Into::into)
}

/// Returns no override when Qoder will choose its account default model at execution time.
#[tauri::command]
pub async fn get_qoder_runtime_config() -> Result<AgentRuntimeConfigResponse, IpcError> {
    tauri::async_runtime::spawn_blocking(|| load_agent_runtime_config(&SystemQoderAdapter))
        .await
        .map_err(|_| IpcError::from(AppError::WorkerFailed))?
        .map(Into::into)
        .map_err(Into::into)
}

/// Sends a bounded query through Qoder's documented stream-json mode.
#[tauri::command]
pub async fn run_qoder_task(request: RunQoderTaskRequest) -> Result<AgentRunResponse, IpcError> {
    tauri::async_runtime::spawn_blocking(move || {
        run_agent_task(&SystemQoderAdapter, &request.query)
    })
    .await
    .map_err(|_| IpcError::from(AppError::WorkerFailed))?
    .map(Into::into)
    .map_err(Into::into)
}
